core/src/main/java/com/linecorp/armeria/common/RequestContextAwareCompletableFuture.java

```diff
@@ -34,213 +37,266 @@
     }
 
     @Override
     public CompletableFuture<T> whenCompleteAsync(BiConsumer<? super T, ? super Throwable> action,
                                                   Executor executor) {
-        return ctx.makeContextAware(super.whenCompleteAsync(ctx.makeContextAware(action), executor));
+        return maybeMakeContextAware(super.whenCompleteAsync(ctx.makeContextAware(action), executor));
     }
 
     @Override
     public <U> CompletableFuture<U> handle(BiFunction<? super T, Throwable, ? extends U> fn) {
-        return ctx.makeContextAware(super.handle(ctx.makeContextAware(fn)));
+        return maybeMakeContextAware(super.handle(ctx.makeContextAware(fn)));
     }
 
     @Override
     public <U> CompletableFuture<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn) {
-        return ctx.makeContextAware(super.handleAsync(ctx.makeContextAware(fn)));
+        return maybeMakeContextAware(super.handleAsync(ctx.makeContextAware(fn)));
     }
 
     @Override
     public <U> CompletableFuture<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn,
                                                 Executor executor) {
-        return ctx.makeContextAware(super.handleAsync(ctx.makeContextAware(fn), executor));
+        return maybeMakeContextAware(super.handleAsync(ctx.makeContextAware(fn), executor));
     }
 
     @Override
     public CompletableFuture<T> exceptionally(Function<Throwable, ? extends T> fn) {
-        return ctx.makeContextAware(super.exceptionally(ctx.makeContextAware(fn)));
+        return maybeMakeContextAware(super.exceptionally(ctx.makeContextAware(fn)));
+    }
+
+    // The methods not available in Java 8 but only in Java 9+
+
+    public <U> CompletableFuture<U> newIncompleteFuture() {
+        return new RequestContextAwareCompletableFuture<>(ctx);
+    }
+
+    public CompletionStage<T> minimalCompletionStage() {
+        return new RequestContextAwareMinimalStage<>(this);
+    }
+
+    public CompletableFuture<T> completeAsync(Supplier<? extends T> supplier) {
+        supplyAsync(this.makeContextAware(supplier)).handle((result, cause) -> {
+            if (cause != null) {
+                completeExceptionally(cause);
+            } else {
+                complete(result);
+            }
+            return null;
+        });
+        return this;
+    }
+
+    public CompletableFuture<T> completeAsync(Supplier<? extends T> supplier,
+                                              Executor executor) {
+        supplyAsync(this.makeContextAware(supplier), executor).handle((result, cause) -> {
+            if (cause != null) {
+                completeExceptionally(cause);
+            } else {
+                complete(result);
+            }
+            return null;
+        });
+        return this;
+    }
+
+    private <T> CompletableFuture<T> maybeMakeContextAware(CompletableFuture<T> future) {
```

Comment:
nit: type `T` is used as a result type for `CompletableFuture<T>`. We should use `<U>` instead.

Reply:
Applied